pages/display.py

```diff
@@ -16,25 +16,48 @@ class Page:
     # Fields stored in DB. See class `pages.models.PageTemplate`
     STORED = ['name', 'h1', 'keywords', 'description', 'title', 'seo_text']
 
-    def __init__(self, page: 'pages.models.Page'=None, context: typing.Dict[str, typing.Any]=None):
+    def __init__(
+        self,
+        context: typing.Dict[str, typing.Any]=None,
+        *,
+        _page: 'pages.models.Page'=None,
+        _name='',
+    ):
         """
         Pass context at ctor, but not render method,
         because client code wants the same context for many different cases.
         """
-        self._page = page
-        self._global_context = defaultdict(dict)
-        if context:
-            self._global_context[page.id] = context
+        self._context = context or {}
+        self._page = _page
+        self._name = _name
+
+    def __set_name__(self, instance, name):
+        # get `Page` instace to create new `Page` instance with `name`
+        page = getattr(instance, name)
+        page_with_name = Page(page._context, _page=page._page, _name=name)
+        # Set new `Page` instance with name
+        setattr(instance, name, page_with_name)
 
     def __get__(self, instance: 'pages.models.Page', type_):
+        if instance and self._name in instance.__dict__:
+            return instance.__dict__[self._name]
         return Page(
-            instance,
-            {'page': instance, **self._global_context[instance.id]}
+            {'page': instance, **self._context},
+            _page=instance,
+            _name=self._name,
         )
 
     def __set__(self, instance: 'pages.models.Page', value: typing.Dict[str, typing.Any]):
-        self._page = instance
-        self._global_context[instance.id] = value
+        if not isinstance(value, dict):
+            raise ValueError(f'Value should be a dict')
+
+        if self._name in instance.__dict__:
+            # merge old context with new one
+            context = {**instance.__dict__[self._name]._context, **value}
+        else:
+            context = {**self._context, **value}
+
+        instance.__dict__[self._name] = Page(context, _page=instance, _name=self._name)
```

Comment:
Let's put value directly to the instance: `instance.__dict__[self._name] = value`
It's pretty the technique from the example from tasks body: https://www.python.org/dev/peps/pep-0487/#trait-descriptors

A code will become much more simple. And we'll be able to throw away not good secret part of the interface

Reply:
`value` has `Dict` type, but `instance.__dict__[self._name]` should be `Page`.
Otherwise `page.display` will return `Dict`, therefore `page.display.h1` will raise an error

Reply:
@artemiy312 , y, it's true. I thought, that py contains some `__getattr__` layer with magic on the top on __dict__.
Even in this case you can use some safe key:
```
# in module display
class Page:
    def __set_name__(self, owner, name):
        self._name = f'_{name}_context_data'
```

`self.key` instead `self._name` looks much more good by the way

Reply:
> Even in this case you can use some safe key:

This code won't work with our descriptor. Each instance will have its own `Page` descriptor instance without `self._name`, so we have to pass it via ctor

>`self.key` instead of `self._name` looks much more good by the way

good point

Reply:
> Each instance will have its own `Page` descriptor instance without `self._name`, so we have to pass it via ctor

seems i don't understand/agree for every of these points. I'll just propose some simpler code in cost of 30 mins and we'll discuss it